### PR TITLE
wip monorepo

### PR DESCRIPTION
**Pull request by rickycodes**
re: #2

### CODE REVIEW

File: .github/workflows/test-monorepo.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  push:
+    branches: main
+  pull_request:
+jobs:
+  checkout_publish:
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          repository: MetaMask/snaps-skunkworks
```
Comment (mcmire) [inline .github/workflows/test-monorepo.yml:12]: Hmm. I'm wary about using real repositories here. I know this action is small, but what we introduce a bug in this action and the dry run doesn't work and we accidentally release a new version? I think if we want to add tests we should have the tests construct their own example repositories. That would also allow us to test edge cases (e.g., what happens if you have a monorepo but one of the workspace package doesn't have a `name` or `version`, etc.). Otherwise maybe it is not worth adding tests, or at least the tests we could ideally have, right now?

**Review by mcmire — COMMENTED** [review]
Left some comments. The implementation seems straightforward enough — I'm just wondering if it's worth adding tests at this point given the concerns I have below.

File: .github/workflows/test-monorepo.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      - uses: actions/setup-node@v3
+        with:
+          node-version: '14'
+      - name: add workspace-tools plugin
+        run: yarn plugin import workspace-tools
+      - name: yarn install
+        run: yarn install
+      - name: Dry Run Publish
+        uses: MetaMask/action-npm-publish@25848c3ea7b4409170bd5bc9b83565852b5a4e42
```
[the remark this answers is not in the document]
Comment (rickycodes) [reply .github/workflows/test-monorepo.yml]: updated to use the relative script (as discussed).

File: scripts/main.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+set -x
+set -e
+set -o pipefail
+
+script_path=$( cd "$(dirname "${BASH_SOURCE[0]}")" ; pwd -P )
+
+"${script_path}"/config_set.sh
```
[the remark this answers is not in the document]
Comment (rickycodes) [reply scripts/main.sh]: it's now just two files and this is happening in `main.sh`

File: scripts/monorepo.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+set -e
+set -o pipefail
+
+script_path=$( cd "$(dirname "${BASH_SOURCE[0]}")" ; pwd -P )
+
+if [[ "$(jq 'has("workspaces")' package.json)" == "true" ]]; then
+  echo "Notice: workspaces detected. Treating as monorepo."
+  yarn workspaces foreach --no-private --verbose exec "${script_path}"/publish.sh
+  exit 0
+fi
```
[the remark this answers is not in the document]
Comment (rickycodes) [reply scripts/monorepo.sh]: updated

File: scripts/main.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+"${script_path}"/monorepo.sh
+
+"${script_path}"/publish.sh
```
[the remark this answers is not in the document]
Comment (rickycodes) [reply scripts/main.sh]: updated

File: test/index.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      if (!error) {
+        reject(new Error('should error when token is invalid'));
+      }
+      t.equal(error.code, 1);
+      resolve();
+    });
+  });
+});
+
+test('should correctly set NPM token', async (t) => {
```
[the remark this answers is not in the document]
Comment (rickycodes) [reply test/index.js]: test removed

File: scripts/publish.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+# Get the published version
+# If the package is not published, set the published version to "NULL"
+NPM_VERSION=$(npm show . version || echo "NULL")
+
+# Get the local version of the package, from package.json
+# The jq "r" flag gives us the raw, unquoted output
+LOCAL_VERSION=$(jq -r .version < package.json)
+
+# Skip Publish if this version is already published
+if [[ $NPM_VERSION != "NULL" || $LOCAL_VERSION == "$NPM_VERSION" ]]; then
```
[the remark this answers is not in the document]
Comment (rickycodes) [reply scripts/publish.sh]: we still need to fallback to something here, but it's just an empty string now and the check is using `-z` as mentioned

File: .github/workflows/test-monorepo.yml (hunk 1 above)
Comment (rickycodes) [reply .github/workflows/test-monorepo.yml:12]: this has been updated to use a ref of a version we know is already published

File: test/index.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+});
+
+test('should only have ten devDependency from now until forever', async (t) => {
+  const { length } = Object.keys(devDependencies);
+  t.equal(length, 10);
+});
+
+test('should not error when performing a dry-run publish', async (t) => {
+  await new Promise((resolve, reject) => {
+    exec('./scripts/publish.sh', (error) => {
```
Comment (mcmire) [inline test/index.js:22]: What if this is `main.sh`, since that is the script that the action will run? Same for the rest of these.

File: test/index.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  });
+});
+
+test('should error when token is invalid', async (t) => {
+  await new Promise((resolve, reject) => {
+    exec(`NPM_TOKEN=${FAKE} ./scripts/publish.sh`, (error) => {
+      if (!error) {
+        reject(new Error('should error when token is invalid'));
+      }
+      t.equal(error.code, 1);
```
Comment (mcmire) [inline test/index.js:50]: Testing the exit code for `main.sh` and not just `publish.sh` is important because we can make sure that the action is reported as a failure if something goes wrong during publishing.

**Review by mcmire — COMMENTED** [review]
One more thing and then I'm good :)

File: test/index.js (hunk 8 above)
Comment (rickycodes) [reply test/index.js]: updated to test `main.sh` instead

File: test/index.js (hunk 9 above)
Comment (rickycodes) [reply test/index.js:50]: updated to test `main.sh` instead

File: scripts/publish.sh [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+NPM_VERSION=$(npm show . version || echo "")
+
+# Get the local version of the package, from package.json
+# The jq "r" flag gives us the raw, unquoted output
+LOCAL_VERSION=$(jq -r .version < package.json)
+
+# Skip Publish if this version is already published
+if [[ -z $NPM_VERSION || $LOCAL_VERSION == "$NPM_VERSION" ]]; then
+  echo "Notice: This module has already been published at this verion. Aborting publish."
+  echo 0
```
Comment (mcmire) [inline scripts/publish.sh:24]: Do we need this line anymore?

**Review by mcmire — COMMENTED** [review]
Sorry one more thing!

File: scripts/publish.sh (hunk 10 above)
Comment (rickycodes) [reply scripts/publish.sh:24]: oh! that's meant to be `exit 0`

Comment (rickycodes) [reply scripts/publish.sh]: updated!

**Review by mcmire — DISMISSED** [review]
LGTM!

**Review by mcmire — APPROVED** [review]
This looks good to me, sorry for the heated debate. Thanks for doing this.

**Review by mcmire — APPROVED** [review]

File: scripts/publish.sh [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+fi
+
+# check param, if it's set (monorepo) we check if it's published before proceeding
+if [[ -n "$1" ]]; then
+  # check if module is published
+  LATEST_VERSION=$(npm view . version || echo "")
+  CURRENT_VERSION=$(jq --raw-output .version package.json)
+  if [ "$LATEST_VERSION" == "$CURRENT_VERSION" ]; then
+    echo "Notice: This module is already published at this version. aborting publish."
+    exit 0
```
Comment (mcmire) [inline scripts/publish.sh:20]: This will not fail the build for a polyrepo. Is that what we want?

File: scripts/main.sh [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+script_path=$( cd "$(dirname "${BASH_SOURCE[0]}")" ; pwd -P )
+
+if [[ -n $NPM_TOKEN ]]; then
+  npm config set //registry.npmjs.org/:_authToken "${NPM_TOKEN}"
+fi
+
+if [[ "$(jq 'has("workspaces")' package.json)" == "true" ]]; then
+  echo "Notice: workspaces detected. Treating as monorepo."
+  yarn workspaces foreach --no-private --verbose exec "$script_path/publish.sh true"
```
Comment (mcmire) [inline scripts/main.sh:15]: Do we want to fail the build when all packages have been published?

File: scripts/publish.sh (hunk 11 above)
Comment (rickycodes) [reply scripts/publish.sh:20]: we're only passing `true` to the publish script for the monorepo case so for polyrepo this whole block will get skipped and things will function as they did before.

Comment (rickycodes) [reply scripts/publish.sh:20]: @mcmire I've actually gone ahead and added two new tests to make this a bit clearer
